src/toil/provisioners/gceProvisioner.py

```diff
@@ -482,56 +487,81 @@ def addNodes(self, nodeType, numNodes, preemptable):
-            if self.config and self.config.sseKey or botoExists:
-                if not self._waitForNode(instance.public_ips[0], 'toil_worker'):
-                    continue
-                retries = 0
-                while True:
-                    try:
-                        if self.config and self.config.sseKey:
-                            self._rsyncNode(instance.public_ips[0], [self.config.sseKey, ':' + self.config.sseKey],
-                                            applianceName='toil_worker')
-                        if botoExists:
-                            self._rsyncNode(instance.public_ips[0], [self.botoPath, ':' + nodeBotoPath],
-                                            applianceName='toil_worker')
-                        break
-                    except Exception as e:
-                        if retries == 5:
-                            log.error("Failed to rysnc keys to worker with ip" % instance.public_ips[0])
-                            log.error('Exception %s', e)
-                            continue # givup
-                        else:
-                            logger.debug("Rsync to new nodes failed, trying again")
-                            retries += 1
-                            time.sleep(20*retries)
-            workersCreated += 1
-        logger.info('Launched %s new instance(s)', numNodes)
-        if len(instancesLaunched) != workersCreated:
-            logger.error("Failed to launch %d workers", len(instancesLaunched)-workersCreated)
+        # Try a few times to create the requested number of workers
+        while numNodes-workersCreated > 0 and retries < 3:
+            instancesLaunched = self.ex_create_multiple_nodes(
+                                    '', nodeType, imageType, numNodes,
+                                    location=self.zone,
+                                    ex_service_accounts=sa_scopes,
+                                    ex_metadata=metadata,
+                                    ex_disks_gce_struct = [disk],
+                                    description=self.tags,
+                                    ex_preemptible = preemptable
+                                    )
+            self.instanceGroup.add_instances(instancesLaunched)
+            failedWorkers = []
+            for instance in instancesLaunched:
+                if self._injectWorkerFiles(instance.public_ips[0]):
+                    workersCreated += 1
+                else:
+                    failedWorkers.append(instance)
+            if failedWorkers:
+                logger.error("Terminating %d failed workers" % len(failedWorkers))
+                self.terminateNodes(failedWorkers)
+            retries += 1
+
+        logger.info('Launched %d new instance(s)', numNodes)
+        if numNodes != workersCreated:
+            logger.error("Failed to launch %d worker(s)", numNodes-workersCreated)
         return workersCreated
 
+
+    def _injectWorkerFiles(self, ip):
```

Comment:
These seem like much more robust ways of dealing with this. Definitely want to keep this model in mind for the provisioner refactor.

Reply:
I was thinking a decorator would look cleaner. Something for the refactor.